experimental/packages/opentelemetry-sdk-metrics-base-wip/src/MeterProvider.ts

```diff
@@ -0,0 +1,165 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as api from '@opentelemetry/api';
+import * as metrics from '@opentelemetry/api-metrics';
+import { Resource } from '@opentelemetry/resources';
+import { Measurement } from './Measurement';
+import { Meter } from './Meter';
+import { Metric } from './Metric';
+import { MetricExporter } from './MetricExporter';
+import { MetricReader } from './MetricReader';
+import { View } from './View';
+
+// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#meterprovider
+
+export type MeterProviderOptions = {
+    resource?: Resource;
+}
+
+export class MeterProvider {
+    private _resource: Resource;
+    private _shutdown = false;
+    private _metricReaders: MetricReader[] = [];
+    private _metricExporters: MetricExporter[] = [];
+    private _views: View[] = [];
+
+    constructor(options: MeterProviderOptions) {
+        this._resource = options.resource ?? Resource.empty();
+    }
+
+    /**
+     * **Unstable**
+     * 
+     * This method is only here to prevent typescript from complaining and may be removed.
+     */
+    getResource() {
+        return this._resource;
+    }
+
+    getMeter(name: string, version = '', options: metrics.MeterOptions = {}): metrics.Meter {
+        // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#meter-creation
+        if (this._shutdown) {
+            api.diag.warn('A shutdown MeterProvider cannot provide a Meter')
+            return metrics.NOOP_METER;
+        }
+
+        // Spec leaves it unspecified if creating a meter with duplicate
+        // name/version returns the same meter. We create a new one here
+        // for simplicity. This may change in the future.
+        return new Meter(this, { name, version }, options.schemaUrl);
+    }
+
+    addMetricReader(metricReader: MetricReader) {
+        this._metricReaders.push(metricReader);
+    }
+
+    addView(view: View) {
+        // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#view
+        this._views.push(view);
+    }
+
+    async shutdown(): Promise<void> {
+        // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#shutdown
+
+        // TODO add a timeout - spec leaves it up the the SDK if this is configurable
+        this._shutdown = true;
```

Comment:
`shutdown()` can only be called once (same as exporter)

Reply:
i'm not sure what change you're suggesting I make here.

Reply:
The spec says

> Shutdown MUST be called only once for each MeterProvider instance.

So something like this?

```ts
// beginning of function
if (this._shutdown) {
  diag.warn("You can only call shutdown() once") // or something to that effect
  return
}
this._shutdown = true

// rest of function ...
```

Reply:
I believe this is specifying the behavior of the MeterProvider, that it only shuts down its child exporters/readers once, but I added a guard.